plugins/army-controller-sanity.cpp

```diff
@@ -0,0 +1,133 @@
+            ac_set.insert(ac);
+        }
+
+        bool ok = true;
+
+        for (auto ent : df::global::world->entities.all)
+        {
+            for (auto ac : ent->army_controllers)
+            {
+                if (ac_set.count(ac) == 0) {
+                    WARN(log).print("acValidationError: Bad controller %p found in entity id %d\n", ac, ent->id);
+                    ok = false;
+                }
+            }
+        }
+
+        for (auto ar : df::global::world->armies.all)
+        {
+            auto ac = ar->controller;
+            if (ac && ac_set.count(ac) == 0) {
+                WARN(log).print("acValidationError: Bad controller %p found in army id %d\n", ac, ar->id);
+                ok = false;
+            }
+            else if (ac && ac->id != ar->controller_id)
+            {
+                WARN(log).print("acValidationError: controller %p id mismatch (%d != %d) in army %d\n", ac, ar->controller_id, ac->id, ar->id);
+                ok = false;
+            }
+            else if (!ac && ar->controller_id != -1)
+            {
+                WARN(log).print("acValidationError: army %d has nonzero controller %d but controller pointer is null\n", ar->id, ar->controller_id);
+                ok = false;
+            }
+        }
+
+        for (auto un : df::global::world->units.all)
+        {
+            auto ac = un->enemy.army_controller;
+            if (ac && ac_set.count(ac) == 0) {
+                WARN(log).print("acValidationError: Bad controller %p found in unit id %d\n", ac, un->id);
+                ok = false;
+            }
+            else if (ac && ac->id != un->enemy.army_controller_id)
+            {
+                WARN(log).print("acValidationError: controller %p id mismatch (%d != %d) in unit %d\n", ac, un->enemy.army_controller_id, ac->id, un->id);
+                ok = false;
+            }
+            else if (!ac && un->enemy.army_controller_id != -1)
+            {
+                WARN(log).print("acValidationError: unit %d has has nonzero controller %d but controller pointer is null\n", un->id, un->enemy.army_controller_id);
+                ok = false;
+            }
+        }
+
+        last_army_controller_next_id = *df::global::army_controller_next_id;
+        last_ac_vec_size = df::global::world->army_controllers.all.size();
+
+        INFO(log).print("acValidation: controller count = %ld, next id = %d, season tick count = %d\n",
+            last_ac_vec_size, last_army_controller_next_id, *df::global::cur_year_tick);
+
+        return ok;
+    }
+}
+
+DFhackCExport DFHack::command_result plugin_init(DFHack::color_ostream& out, std::vector <DFHack::PluginCommand>& commands)
+{
+    return DFHack::CR_OK;
+}
+
+DFhackCExport DFHack::command_result plugin_enable(DFHack::color_ostream& out, bool enable)
+{
+    is_enabled = enable;
+    return DFHack::CR_OK;
+}
+
+DFhackCExport DFHack::command_result plugin_onupdate(DFHack::color_ostream& out)
+{
+    if (is_enabled)
```

Comment:
This isn't required. `plugin_onupdate` only gets called if `is_enabled` is true

Reply:
however, it might be a good idea to only run once per tick.

Reply:
the cost to test if it's running only once a tick is basically as expensive as the test that checks if the army controller list length has changed, so i don't see any benefit to it